distributed/shuffle/_scheduler_plugin.py

```diff
@@ -309,34 +325,60 @@ def _unset_restriction(self, ts: TaskState) -> None:
         original_restrictions = ts.annotations.pop("shuffle_original_restrictions")
         self.scheduler.set_restrictions({ts.key: original_restrictions})
 
+    def _restart_recommendations(self, id: ShuffleId) -> Recs:
+        barrier_task = self.scheduler.tasks[barrier_key(id)]
+        recs: Recs = {}
+
+        for dt in barrier_task.dependents:
+            if dt.state == "erred":
+                return {}
+            recs.update({dt.key: "released"})
+
+        if barrier_task.state == "erred":
+            return {}  # pragma: no cover
```

Comment:
why don't you want coverage to detect this? Seems like an important case

Reply:
https://github.com/dask/distributed/pull/7970#discussion_r1263396997

Reply:
Added a comment to explain

Reply:
This seems like an ideal case for an `assert False, "Invariant broken"` ?

Reply:
That would also work. I'm wondering if `assert False` is the right thing to add here given that `PYTHONOPTIMIZE` will strip them. It would work as an addition though.

Reply:
raising a `RuntimeError` now